internal/controller/controllers/agent_controller.go

```diff
@@ -77,69 +94,208 @@ func (r *AgentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-			clientError = false
-		}
-		clusterDeploymentRefErr := newKubeAPIError(errors.Wrapf(err, errMsg), clientError)
 		// Update that we failed to retrieve the clusterDeployment
-		r.updateFailure(ctx, agent, clusterDeploymentRefErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, errors.Wrapf(err, errMsg), !k8serrors.IsNotFound(err))
 	}
 
 	// Retrieve cluster for ClusterDeploymentName from the database
 	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
 	if err != nil {
 		if gorm.IsRecordNotFoundError(err) {
-			Requeue = true
 			inventoryErr = common.NewApiError(http.StatusNotFound, err)
+			internalError = false
 		} else {
-			Requeue = false
 			inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+			internalError = true
 		}
 		// Update that we failed to retrieve the cluster from the database
-		r.updateFailure(ctx, agent, inventoryErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, inventoryErr, internalError)
+	}
+
+	//Retrieve host from cluster
+	host := getHostFromCluster(cluster, agent.Name)
+	if host == nil {
+		return r.updateStatus(ctx, agent, host, errors.New("host not found in cluster"), false)
 	}
 
-	var result ctrl.Result
 	// check for updates from user, compare spec and update if needed
-	result, err = r.updateIfNeeded(ctx, agent, cluster)
+	err = r.updateIfNeeded(ctx, agent, cluster)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return result, nil
+		return r.updateStatus(ctx, agent, host, err, !IsHTTP4XXError(err))
 	}
 
-	err = r.updateInventory(cluster, agent)
+	err = r.updateInventory(host, agent)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return ctrl.Result{Requeue: true}, nil
+		return r.updateStatus(ctx, agent, host, err, true)
 	}
 
-	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
-		Type:    adiiov1alpha1.AgentSyncedCondition,
-		Status:  corev1.ConditionTrue,
-		Reason:  adiiov1alpha1.AgentSyncedReason,
-		Message: adiiov1alpha1.AgentStateSynced,
-	})
+	return r.updateStatus(ctx, agent, host, nil, false)
+}
+
+func (r *AgentReconciler) updateStatus(ctx context.Context, agent *adiiov1alpha1.Agent, h *models.Host, syncErr error, internal bool) (ctrl.Result, error) {
+
+	r.specSynced(agent, syncErr, internal)
+
+	if h != nil && h.Status != nil {
+		status := *h.Status
+
+		r.connected(agent, status)
+		r.readyForInstallation(agent, status)
+		r.validated(agent, status, h)
+		r.installed(agent, status, swag.StringValue(h.StatusInfo))
+	}
 	if updateErr := r.Status().Update(ctx, agent); updateErr != nil {
-		r.Log.WithError(updateErr).Error("failed to update agent status")
-		return ctrl.Result{Requeue: true}, nil
+		r.Log.WithError(updateErr).Error("failed to update agent Status")
+		return ctrl.Result{Requeue: true}, updateErr
 	}
-	return result, nil
+	if internal {
+		return ctrl.Result{Requeue: true, RequeueAfter: defaultRequeueAfterOnError}, nil
```

Comment:
Only one of these should be specified. Either you want to requeue immediatly, or you want to requeue later.

Reply:
Done